File: typescript/src/repositories/creator-repository.ts
import type { QueryResultRow } from "pg";
import type { Creator } from "../data/models.js";
import type { PoolOrClient } from "../db/pool.js";

function toCreator(row: QueryResultRow): Creator {
  return {
    id: row.id,
    displayName: row.display_name,
    email: row.email,
    balanceCents: Number(row.balance_cents),
    createdAt: row.created_at.toISOString(),
  };
}

export class CreatorRepository {
  constructor(private db: PoolOrClient) {}

  async findById(id: string): Promise<Creator | null> {
    const { rows } = await this.db.query(
      "SELECT * FROM creators WHERE id = $1",
      [id],
    );
    return rows.length > 0 ? toCreator(rows[0]) : null;
  }

  async updateBalance(id: string, balanceCents: number): Promise<void> {
    await this.db.query(
      "UPDATE creators SET balance_cents = $1 WHERE id = $2",
      [balanceCents, id],
    );
  }
}

File: python/src/repositories/creator_repository.py
import asyncpg

from src.data.models import Creator


def _to_creator(row: asyncpg.Record) -> Creator:
    return Creator(
        id=row["id"],
        display_name=row["display_name"],
        email=row["email"],
        balance_cents=row["balance_cents"],
        created_at=row["created_at"].isoformat(),
    )


class CreatorRepository:
    def __init__(self, db: asyncpg.Pool | asyncpg.Connection):
        self.db = db

    async def find_by_id(self, id: str) -> Creator | None:
        row = await self.db.fetchrow("SELECT * FROM creators WHERE id = $1", id)
        return _to_creator(row) if row else None

    async def update_balance(self, id: str, balance_cents: int) -> None:
        await self.db.execute(
            "UPDATE creators SET balance_cents = $1 WHERE id = $2", balance_cents, id
        )
